core/shared/src/main/scala/fs2/concurrent/Signal.scala

```diff
@@ -390,31 +390,65 @@ trait SignallingMapRef[F[_], K, V] extends MapRef[F, K, V] {
 object SignallingMapRef {
 
   /** Builds a `SignallingMapRef` for effect `F`, initialized to the supplied value.
+    *
+    * Update semantics for `discrete` are the same as `SignallingRef`, with one exception:
+    * it cannot distinguish updates that remove a key (by setting its value to `None`).
+    *
+    * More specifically: if you remove a key, this will only notify once per listener
+    * i.e. setting it to `None` again will not trigger another update.
+    * Furthermore, if a listener's last pull returned `None`, and by the time it pulls again the
+    * current value is `None`, then it will not be notified regardless of any non-`None` updates
+    * that may have happened between the pulls. This special semantic for `None` is necessary to
+    * prevent memory leaks at keys with no values and no listeners.
     */
   def ofSingleImmutableMap[F[_], K, V](
       initial: Map[K, V] = Map.empty[K, V]
   )(implicit F: Concurrent[F]): F[SignallingMapRef[F, K, Option[V]]] = {
     case class State(
-        value: Map[K, V],
         lastUpdate: Long,
-        listeners: Map[K, LongMap[Deferred[F, (Option[V], Long)]]]
+        keys: Map[K, KeyState]
     )
 
-    F.ref(State(initial, 0L, initial.flatMap(_ => Nil)))
+    type Listener = Deferred[F, (Option[V], Long)]
+
+    // fix lastUpdate at -1 when the value is not present
+    // so we don't have to keep an update counter when the key has neither a value nor any listeners
+    case class KeyState(
+        value: Option[V],
+        lastUpdate: Long,
+        listeners: LongMap[Listener]
+    )
+
+    F.ref(State(0L, initial.map { case (k, v) => k -> KeyState(Some(v), 0L, LongMap.empty) }))
       .product(F.ref(1L))
       .map { case (state, ids) =>
         def newId = ids.getAndUpdate(_ + 1)
 
         def updateAndNotify[U](state: State, k: K, f: Option[V] => (Option[V], U))
             : (State, F[U]) = {
-          val (newValue, result) = f(state.value.get(k))
-          val newMap = newValue.fold(state.value - k)(v => state.value + (k -> v))
-          val lastUpdate = state.lastUpdate + 1
-          val newListeners = state.listeners - k
-          val newState = State(newMap, lastUpdate, newListeners)
-          val notifyListeners = state.listeners.get(k).fold(F.unit) { listeners =>
-            listeners.values.toVector.traverse_ { listener =>
-              listener.complete(newValue -> lastUpdate)
+
+          val keyState = state.keys.get(k)
+
+          val (newValue, result) = f(keyState.flatMap(_.value))
+
+          val lastUpdate = {
+            val lu = state.lastUpdate + 1
+            // skip -1 b/c of its special semantic
+            if (lu == -1L) 0L else lu
```

Comment:
Sorry if it's a stupid question, but should we increment `lu` (`state.lastUpdate + 1`) before checking for `-1`? 🤔 
Also, can we ever get `lastUpdate == -1` in the overarching `State` (not in the `KeyState` for each key)? 🙂

Reply:
> but should we increment `lu` (`state.lastUpdate + 1`) before checking for `-1`?

Yes, if we increment and it's `-1` then we want to skip. If the current value is `-1` then we've already failed 😉 



> Also, can we ever get `lastUpdate == -1` in the overarching `State` (not in the `KeyState` for each key)?

Every update causes `state.lastUpdate` to increment. Additionally, `state.keys(k).lastUpdate` takes the same value when `k` is updated.

Reply:
Thanks, I think I understand it now. 👍